### PR TITLE
Ports - Ari

### PR DESCRIPTION
**Pull request by aribray**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | My top_ten method returns an array of Works based on their category. It sorts the results in descending order based on votes. If no works in the category have any votes, it just returns an array of 10 works based on category.
Describe how you approached testing that model method. What edge cases did you come up with? | I had a hard time testing this method. When I first created votes, I used a gem (acts_as_votable) that made it really easy to test. I refactored my votes model to be less...extra, and I lost a lot of the functionality that made testing really easy. I had a hard time getting the votes in my YAML file to stick to the works in my YAML file. So some edge case tests that were working initially don't work anymore. One of my tests expected that if any works in a category had votes, there would be no works with 0 votes in the top ten for that category. 
What are `session` and `flash`? What is the difference between them? | session and flash are both hash-like objects. session's contents persist through the whole, well, session. flash's contents persist only until the next HTTP request.
What was one thing that you gained more clarity on through this assignment? | I learned more about the relationships between tables (and how tricky it can be to test those).
What is the Heroku URL of your deployed application? | http://ada-mediaranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | Throw fixtures into the fire.

### CODE REVIEW

**Comment by tildeee** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | x
Comprehension questions | x, oops :fire:
**General** |
Rails fundamentals (RESTful routing, use of named paths) | x
Views are well-organized (DRY, use of semantic HTML, use of partials) | x, opportunity to refactor the `main` and `works#index` pages with partials
Errors are reported to the user | x
Business logic lives in the models | x
Models are thoroughly tested, including relations, validations and any custom logic | x
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work | x
**Wave 1 - Media** |
Splash page shows the three media categories | x
Basic CRUD operations on media are present and functional | x
**Wave 2 - Users and Votes** |
Users can log in and log out | x
The ID of the current user is stored in the session | x
A user cannot vote for the same media more than once | x
All media lists are ordered by vote count | x
Splash page contains a media spotlight | x
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | x
Individual user pages and the user list are present | x
**Optional - Styling** |
Bootstrap is used appropriately | x
Look and feel is similar to the original | x
**Overall** |



Note: I graded this after you added OAuth (which makes me vvvv happy :) :) :) )


Well done on this project, Ari! Overall, you hit all of the learning goals, you did all of the requirements, and your feature code and tests look good. Well done-- this was a huge project.


I have a few comments on opportunities for refactoring.


That being said-- good work!

File: app/models/work.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  validates :title, presence: true
+
+  has_many :votes
+  has_many :users, through: :votes
+
+  def self.top_ten(media)
+    results = Work.where(media: media.to_s)
+    results = results.sort_by {|result| result.votes.length}.reverse
+
+    results = Work.where(media: media.to_s) if results.empty?
```
Comment (tildeee) [inline app/models/work.rb:13]: What does this line do? Do we need it? Commenting it out still lets the tests pass

File: app/models/work.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    if results.length <= 10
+      return results
+    else
+      results = results[0..9]
```
Comment (tildeee) [inline app/models/work.rb:18]: In Ruby, slicing the array with `[0..9]` actually works out in your favor. Consider that taking out the `if ... else` and just doing `return results[0..9]` makes your tests pass

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    end
+
+    results
+  end
```
Comment (tildeee) [inline app/models/work.rb:22]: Consider this refactoring of this method:

```ruby
  def self.top_ten(media)
    results = Work.where(media: media)
    results = results.sort_by { |result| result.votes.length }.reverse
    return results[0..9]
  end
```

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Work < ApplicationRecord
+  validates :title, presence: true
+
+  has_many :votes
+  has_many :users, through: :votes
+
+  def self.top_ten(media)
+    results = Work.where(media: media.to_s)
```
Comment (tildeee) [inline app/models/work.rb:10]: Hm... do you need to do `media.to_s`?

File: app/models/work.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      results = results[0..9]
+    end
+
+    results
+  end
+
+  def self.spotlight
+    works = Work.all
+    spotlight = works.sort_by {|work| work.votes.length}.reverse
+    spotlight.first
```
Comment (tildeee) [inline app/models/work.rb:27]: You know I prefer an explicit `return` ;) Also, as a future refactoring, it may be fine at this point to just inline `works = Work.all`, so just `spotlight = Work.all.sort_by { |work| work.votes.length }.reverse`

File: app/views/works/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,94 @@
+<h2>List of Works </h2>
+
+<section class="album">
+  <h3>Albums</h3>
+  <% albums = Work.where(media: "album")%>
+  <% albums = albums.sort_by {|work| work.votes.length}.reverse%>
```
Comment (tildeee) [inline app/views/works/index.html.erb:6]: It may make more sense to put this logic in the controller, and then the eventual value of `albums` as an instance variable `@albums`

Comment (tildeee) [reply app/views/works/index.html.erb:6]: You defined the method `top_ten` in your `Work` model, but didn't use it here! :O

File: config/routes.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+Rails.application.routes.draw do
+  root to: 'works#main'
+  resources :works
+
+  get '/auth/github', as: 'github_login'
+  get '/auth/:provider/callback', to: 'users#create'
+  delete '/logout', to: 'users#destroy', as: 'logout'
+  get '/users', to: 'users#index', as: 'users'
+
+  # combine current user with users#show?
```
Comment (tildeee) [inline config/routes.rb:12]: GOOD CALL!

File: test/models/work_test.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  describe 'relationships' do
+    it 'has many votes' do
+      album.votes << vote1
+      album.votes << vote2
+      expect(album.votes.length).must_equal 2
+    end
+  end
+
+  describe 'top_10 method' do
+    it 'will return up to 10 works (if no works have upvotes, will return 10 works)' do
```
Comment (tildeee) [inline test/models/work_test.rb:39]: Hm... seems like your test description and test don't match?
